### PR TITLE
Parser: Unary minus

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implement unary minus.

https://www.pivotaltracker.com/story/show/182497332

### Important Notes

- This one had a lot of edge cases, so it has a lot of tests.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

[ci no changelog needed]

### CODE REVIEW

File: lib/rust/parser/tests/parse.rs [hunk 1]
```diff
@@ -361,6 +361,55 @@ fn plus_negative() {
     test(&code.join("\n"), expected);
 }
 
+#[test]
+fn minus_binary() {
+    let cases = [
+        ("x - 5", block![(OprApp (Ident x) (Ok "-") (Number 5))]),
+        ("x-5", block![(OprApp (Ident x) (Ok "-") (Number 5))]),
+        ("x.-y", block![(OprApp (Ident x) (Ok ".") (UnaryOprApp "-" (Ident y)))]),
```
Comment (JaroslavTulach) [inline lib/rust/parser/tests/parse.rs:383]: @jdunkerley mentioned to me that he'd like to see `-7.absolute` to evaluate to `7` and not `-7` as is currently the case...

File: lib/rust/parser/tests/parse.rs [hunk 2]
```diff
@@ -361,6 +361,55 @@ fn plus_negative() {
...
+fn minus_unary() {
+    #[rustfmt::skip]
+    let cases = [
+        ("f -5", block![(App (Ident f) (UnaryOprApp "-" (Number 5)))]),
+        ("-5", block![(UnaryOprApp "-" (Number 5))]),
+        ("(-5)", block![
+            (MultiSegmentApp #(((Symbol "(") (UnaryOprApp "-" (Number 5))) ((Symbol ")") ())))]),
+        ("-(x * 2)", block![
+            (UnaryOprApp "-" (MultiSegmentApp #(
+             ((Symbol "(") (OprApp (Ident x) (Ok "*") (Number 2)))
```
Comment (JaroslavTulach) [inline lib/rust/parser/tests/parse.rs]: Out of curiosity. There is `Symbol "("` in the tree. Shouldn't parenthesis disappear and be replaced by the structure of the tree?

**Review by JaroslavTulach — APPROVED** [review]
My minimal knowledge of Rust doesn't qualify me as a real reviewer, I am afraid. However, as far as I can say the tests look like a step towards right direction. Two notes added...

File: lib/rust/parser/tests/parse.rs (hunk 1 above)
Comment (kazcw) [reply lib/rust/parser/tests/parse.rs:383]: Ok. I have updated the test case for that (see the diff in `minus_unary`), and bumped the precedence of unary `-` so that it binds more tightly than binary `.` (and the changed test passes)

File: lib/rust/parser/tests/parse.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/tests/parse.rs]: We need every token represented somewhere, but `MultiSegmentApp` is an obtuse way to do that for parentheses; I am planning to introduce a node type for parenthesized expressions; I might be able to fit that in today.

File: lib/rust/parser/src/syntax/operator.rs [hunk 3]
```diff
@@ -70,110 +70,180 @@ pub fn resolve_operator_precedence_if_non_empty<'s>(
...
+            expression.operand(item);
+        }
+    }
+    expression.finish()
+}
+
+
+// === Expression builder ===
+
+/// Stack machine that builds an expression from syntax nodes according to a shift-reduce algorithm.
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs]: put some more info about the algorithm .Maybe a link to shuffling-yard would be useful here too :)

File: lib/rust/parser/src/syntax/operator.rs [hunk 4]
```diff
@@ -70,110 +70,180 @@ pub fn resolve_operator_precedence_if_non_empty<'s>(
...
+            (false, None, Some(_)) => self.operand(syntax::Tree::unary_opr_app(opr, None).into()),
+            // Binary operator section (no LHS).
+            (_, Some(prec), _) => self.binary_operator(prec, opr),
+            // Failed to compute a role for the operator; this should not be possible.
+            (_, None, None) => {
+                // We don't know the correct precedence, so we can't structure the tree correctly.
+                // Pick some arbitrary value so we can at least produce *some* tree containing all
+                // the right subexpressions; we'll wrap the expression in an `Invalid` node.
+                const ARBITRARY_PRECEDENCE: usize = 20;
+                let error = || format!("Precedence of: {:?}", opr.code);
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs]: I love the comments here!

**Review by wdanilo — APPROVED** [review]
